src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -278,203 +289,238 @@ pretty_print_platform_info(const boost::property_tree::ptree& _ptDevice)
-      if(!dsa.hasFlashImage)
-        throw xrt_core::error("Flash image is not available");
-      boardsToUpdate.push_back(std::make_pair(device->get_device_id(), dsa));
-    }
+  // If the shell is not up-to-date and dsa has a flash image, queue the board for update
+  bool same_shell = pt.get<bool>(std::to_string(workingDevice->get_device_id()) + ".platform.status.shell");
+  bool same_sc = pt.get<bool>(std::to_string(workingDevice->get_device_id()) + ".platform.status.sc");
+
+  // Always update Arista devices
+  auto vendor = xrt_core::device_query<xrt_core::query::pcie_vendor>(workingDevice);
+  if (vendor == ARISTA_ID)
+    same_shell = false;
+
+  if (XBU::getForce()) {
+    same_shell = false;
+    same_sc = false;
+  }
+
+  if (!same_shell || !same_sc) {
+    if(!dsa.hasFlashImage)
+      throw xrt_core::error("Flash image is not available");
+
+    boardsToUpdate.push_back(std::make_pair(workingDevice->get_device_id(), dsa));
+  }
+  
+  // Is there anything to flash
+  if (boardsToUpdate.empty() == true) {
+    std::cout << "\nDevice is up-to-date.  No flashing to performed.\n";
+    return;
   }
 
   // Continue to flash whatever we have collected in boardsToUpdate.
-  uint16_t success = 0;
   bool needreboot = false;
   bool need_warm_reboot = false;
-  std::stringstream report_status;
-  if (!boardsToUpdate.empty()) {
+  std::stringstream report_stream;
 
-    // Prompt user about what boards will be updated and ask for permission.
-    if(!XBU::can_proceed(XBU::getForce()))
-      return;
+  // Prompt user about what boards will be updated and ask for permission.
+  if(!XBU::can_proceed(XBU::getForce()))
+    return;
 
-    // Perform DSA and BMC updating
-    for (auto& p : boardsToUpdate) {
-      bool reboot = false;
-      bool warm_reboot = false;
+  // Perform DSA and BMC updating
+  std::stringstream error_stream;
+  for (auto& p : boardsToUpdate) {
+    try {
       std::cout << std::endl;
-      try {
-        updateShellAndSC(p.first, p.second, reboot, warm_reboot);
-        report_status << boost::format("  [%s] : Successfully flashed\n") % getBDF(p.first);
-        success++;
+      // 1) Flash the Satellite Controller image
+      if (update_sc(p.first, p.second) == true)  {
+        report_stream << boost::format("  [%s] : Successfully flashed the Satellite Controller (SC) image\n") % getBDF(p.first);
+        need_warm_reboot = true;
+      } else
+        report_stream << boost::format("  [%s] : Satellite Controller (SC) is either up-to-date, fixed, or not installed. No actions taken.\n") % getBDF(p.first);
+
+      // 2) Flash shell image
+      if (update_shell(p.first, p.second) == true)  {
+        report_stream << boost::format("  [%s] : Successfully flashed the base (e.g., shell) image\n") % getBDF(p.first);
+        needreboot = true;
+      } else
+        report_stream << boost::format("  [%s] : Base (e.g., shell) image is up-to-date.  No actions taken.\n") % getBDF(p.first);
```

Comment:
Should SC flash be independent of shell flash, i.e., have their own try/catches. I think we should continue to try and flash the shell even if the SC flash fails (eg: when an SC is not operational)

Reply:
Hmm...is this a requirement that Max or Sonal has mentioned.   

My logic behind this code is simple:
_If the SC flash failed then we shouldn't attempt to program the shell._

For if we do, then we could brick the board even further.

Note: Failure to program the board is NOT the same as skipping programing the SC (e.g., fixed or versions the same).

Thoughts?

Reply:
As far as I know, it is not a requirement. But as per the old flashing code, we would give a warning:
`WARNING: Failed to update SC firmware on card [...]`
and continue with flashing the shell.

@maxzhen, what is the desired behavior here?

Reply:
In the early days, SC flash fails left and right and board works just fine without SC. Hence we don't really care if SC flashed properly back then. I think we should stop updating the board if SC can't be flash now since if we continue, the board will be in a unsupported state.
I'm merging this PR now since @AShivangi has already approved the change (but forgot to update the status)